### PR TITLE
sql,cli: improve statement bundles for prepared statements

### LINKED ISSUE #114356: sql: change format of statement.sql in statement diagnostics bundles for prepared statements
When capturing a statement diagnostics bundle for a prepared statement, the statement.sql file has the format:

```sql
INSERT INTO foo VALUES ($1, $2, $3)

-- Arguments:
--  $1: NULL
--  $2: 100
--  $3: 'xyz'
```

Every time we want to use this statement we need to manually turn it into:

```sql
PREPARE stmt AS INSERT INTO foo VALUES ($1, $2, $3);
EXECUTE stmt (NULL, 100, 'xyz');
```

Or even this if we don't want to go through prepared statements:

```sql
INSERT INTO foo VALUES (NULL, 100, 'xyz');
```

It would be great if we could use the PREPARE / EXECUTE format for statement.sql instead of what we have now. Or, if we want to keep statement.sql the way it is, also print to another file or two in these two formats. Would save a lot of time.

Jira issue: CRDB-33476
Epic CRDB-34181

### PR DESCRIPTION
**Pull request by yuzefovich**
This commits adds a couple of new files to the stmt bundles that were collected for prepared statements:
- `statement-prepared.sql` contains `PREPARE` and `EXECUTE` statements that can be copy-pasted to reproduce how the original statement was executed
- `statement-no-placeholders.sql` contains the original statement in which all placeholders were replaces with their values.

Additionally, `debug statement-bundle recreate` command has been taught to parse `statement.sql` and replace placeholders with their values before printing out the statement into the sql shell (which is the last step of the recreating the bundle). This approach will work even on bundles that were collected on older binaries without this change.

These improvements should make debugging prepared statementsa bit easier.

Fixes: #114356.
Epic: CRDB-34181.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
Actually, hold off for a sec - I'll try to add a test for this.

**Comment by michae2** [comment]
You could probably mark this as fixing https://github.com/cockroachdb/cockroach/issues/114356

**Review by yuzefovich — COMMENTED** [review]
Indeed, thanks for the callout. Added a test, RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-Np1w77g93Oidlps5I_A:bcf4gvb)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

**Comment by yuzefovich** [comment]
Hm, actually I wonder whether we should do what you suggested in #114356 - namely, introduce another file into the stmt bundle (perhaps named `statement-no-placeholders.sql`) that will be generated every time when the prepared statements were used and would contain this "filled in" statement. Then, in `statement-bundle recreate` we'll look at contents of the bundle and will pick `statement-no-placeholders.sql` if it exists. Maybe we'd keep the logic for "filling in placeholder values" in `bundle recreate` command in case we're recreating a bundle collected on a binary without `statement-no-placeholders.sql`? Thoughts?

**Comment by michae2** [comment]
> Hm, actually I wonder whether we should do what you suggested in #114356 - namely, introduce another file into the stmt bundle (perhaps named `statement-no-placeholders.sql`) that will be generated every time when the prepared statements were used and would contain this "filled in" statement. Then, in `statement-bundle recreate` we'll look at contents of the bundle and will pick `statement-no-placeholders.sql` if it exists. Maybe we'd keep the logic for "filling in placeholder values" in `bundle recreate` command in case we're recreating a bundle collected on a binary without `statement-no-placeholders.sql`? Thoughts?

Up to you, I'm happy either way! 🙂

**Review by yuzefovich — COMMENTED** [review]
I decided to add the new file to the bundle, RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-Np288ti2oAZwA4fe4RX:b9grf9y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

**Review by michae2 — APPROVED** [review]
:lgtm: Nice!

> Reviewed 1 of 2 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-Np653ws6sqce2LXQmXI:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/explain_bundle.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-Np5w-BFBkGKLWbyw6ne:-Np5w-BFBkGKLWbyw6nf:b-8hq1om) ([raw file](https://github.com/cockroachdb/cockroach/blob/81df71e9c323efa6f05316889e505256ce2cdab2/pkg/sql/explain_bundle.go#L349)):*
> ```Go
> 	if b.placeholders != nil && len(b.placeholders.Values) != 0 {
> 		var buf bytes.Buffer
> 		buf.WriteString(output)
> ```

idea: I'm wondering, while you're here, what would you think of adding `buf.WriteString("PREPARE p AS ")` before this, and then changing the arguments part to something like:

```go
buf.WriteString("EXECUTE p (")
for i, v := range b.placeholders.Values {
  if b.flags.RedactValues {
    fmt.Fprintf(&buf, "\n  /* %s */ %s", tree.PlaceholderIdx(i), redact.RedactedMarker())
  } else {
    fmt.Fprintf(&buf, "\n  /* %s */ %v", tree.PlaceholderIdx(i), v)
  }
  if i < len() - 1 {
    fmt.Fprintf(",")
  }
}
buf.WriteString("\n);\n")
```

That would make it easier to copy and paste the prepared statement version, in case we also want to use that.

I know you already wrote a beautiful `ReplacePlaceholdersWithValuesForBundle` for parsing the old format, and I think we could keep that, too, at least in bundle recreate, since we'll still have lots of bundles using the old format for years to come.

**Review by yuzefovich — COMMENTED** [review]
Thanks for the suggestions and the review! I see that Drew also typed something - I'll wait for his feedback before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-Np78XHN5_jZTc0C4-sM:bz0ohxu)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/explain_bundle.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-Np5w-BFBkGKLWbyw6ne:-Np77iKPC-jY3lciTmFW:bip3oqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/81df71e9c323efa6f05316889e505256ce2cdab2/pkg/sql/explain_bundle.go#L349)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

idea: I'm wondering, while you're here, what would you think of adding `buf.WriteString("PREPARE p AS ")` before this, and then changing the arguments part to something like:

```go
buf.WriteString("EXECUTE p (")
for i, v := range b.placeholders.Values {
  if b.flags.RedactValues {
    fmt.Fprintf(&buf, "\n  /* %s */ %s", tree.PlaceholderIdx(i), redact.RedactedMarker())
  } else {
    fmt.Fprintf(&buf, "\n  /* %s */ %v", tree.PlaceholderIdx(i), v)
  }
  if i < len() - 1 {
    fmt.Fprintf(",")
  }
}
buf.WriteString("\n);\n")
```

That would make it easier to copy and paste the prepared statement version, in case we also want to use that.

I know you already wrote a beautiful `ReplacePlaceholdersWithValuesForBundle` for parsing the old format, and I think we could keep that, too, at least in bundle recreate, since we'll still have lots of bundles using the old format for years to come.
</blockquote></details>

That's a good idea. I think it's worth keeping the original format of `statement.sql` (so that `debug statement-bundle recreate` could parse it more easily), so I added another file `statement-prepared.sql` which contains `PREPARE` and `EXECUTE` statements like you suggested.

**Review by DrewKimball — APPROVED** [review]
Sorry for the delay, :lgtm:!

> Reviewed 1 of 2 files at r2, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-NpLOPo4-fJLJTLSCEK_:b3y7t8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)

**Review by michae2 — COMMENTED** [review]
This is going to improve our quality of life so much!!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-:-NpLPyDd5PnQg7ZjkCs4:b-2bgeod)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/sql/explain_bundle.go` line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118316#-Np5w-BFBkGKLWbyw6ne:-NpLPLSG48mCZM9QYuCT:bee8dxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/81df71e9c323efa6f05316889e505256ce2cdab2/pkg/sql/explain_bundle.go#L349)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

That's a good idea. I think it's worth keeping the original format of `statement.sql` (so that `debug statement-bundle recreate` could parse it more easily), so I added another file `statement-prepared.sql` which contains `PREPARE` and `EXECUTE` statements like you suggested.
</blockquote></details>

Thank you! :heart:

**Pull request merged**
